rand_core/src/impls.rs

```diff
@@ -72,32 +75,35 @@ impl Observable for u64 {
 /// Returns `(n, byte_len)`. `src[..n]` is consumed (and possibly mutated),
 /// `dest[..byte_len]` is filled. `src[n..]` and `dest[byte_len..]` are left
 /// unaltered.
-fn fill_via_chunks<T: Observable>(src: &mut [T], dest: &mut [u8]) -> (usize, usize) {
+fn fill_via_chunks<T: Observable>(src: &[T], dest: &mut [u8]) -> (usize, usize) {
     let size = core::mem::size_of::<T>();
-    let byte_len = min(core::mem::size_of_val(src), dest.len());
-    let num_chunks = (byte_len + size - 1) / size;
-
-    // Byte-swap for portability of results. This must happen before copying
-    // since the size of dest is not guaranteed to be a multiple of T or to be
-    // sufficiently aligned.
-    if cfg!(target_endian = "big") {
-        for x in &mut src[..num_chunks] {
-            *x = x.to_le();
-        }
-    }
 
-    dest[..byte_len].copy_from_slice(&<[T]>::as_bytes(&src[..num_chunks])[..byte_len]);
+    // Always use little endian for portability of results.
 
-    (num_chunks, byte_len)
+    let mut dest = dest.chunks_exact_mut(size);
+    let mut src = src.iter();
+
+    let zipped = dest.by_ref().zip(src.by_ref());
+    let num_chunks = zipped.len();
+    zipped.for_each(|(dest, src)| dest.copy_from_slice(src.to_le_bytes().as_ref()));
```

Comment:
Nit: It seems like this crate generally prefers `for` loops to `for_each`, is there a reason to use `for_each` here? Is it because of variable shadowing?

Reply:
The compiler doesn't optimise the same for `for` loops vs `for_each`, and for ensuring certain optimisations take place, you need to use `for_each`. In like 90% of cases, `for` loops are just fine, but when targeting particular optimisation cases, `for_each` is usually what you need to reach for, particularly when you are trying to get autovectorisation to kick in.

Reply:
For the record, I have nothing against usage of `for_each` in this crate.

Reply:
Just as a reference: https://rust.godbolt.org/z/3qfned8vn

This shows the compiled output diff between the `for` loop and `for_each`. The `for_each` is able to optimise to the two `memcpy` ops, whereas the `for` optimises to a single `memcpy` case with a separate path that does not have `memcpy`.

Reply:
Huh. Do we want the tail-case to use memcpy? In the worst case it's only 7 bytes

Reply:
Looking at the asm linked above, it seems like the `for` loop gets compiled to a load-store loop while `for_each` gets compiled into a single `memcpy`. In both cases, the tail-case uses a memcpy. The tail case using a memcpy is to be expected, and seems fine to just defer to LLVM here.

So it seems like keeping the `for_each` is fine.